### ISSUE TITLE
Installed packages are not available in produced image (alpine/apk/bash)

### ISSUE THREAD

**Issue by very-doge-wow** [GoogleContainerTools/kaniko#2177]
**Actual behavior**
I'm using Kaniko v1.8.1 to build an image using a Dockerfile.
I could reproduce the issue with a minimal Dockerfile, which looks like this:
```Dockerfile
FROM alpine:3.16

USER root

RUN apk add --no-cache bash git

USER 1000

SHELL ["/bin/bash", "-c"]

```

The image is built correctly and judging by the logoutput, the packages get installed correctly (I've removed company urls/mirrors):
```
INFO[0000] Retrieving image manifest XXX/alpine:3.16 
INFO[0000] Retrieving image XXX/alpine:3.16 from registry XXX 
INFO[0000] Built cross stage deps: map[]                
INFO[0000] Retrieving image manifest XXX/alpine:3.16 
INFO[0000] Returning cached image manifest              
INFO[0000] Executing 0 build triggers                   
INFO[0000] Unpacking rootfs as cmd RUN apk add --no-cache bash requires it. 
INFO[0000] USER root                                    
INFO[0000] cmd: USER                                    
INFO[0000] RUN apk add --no-cache bash                  
INFO[0000] Taking snapshot of full filesystem...        
INFO[0000] cmd: /bin/sh                                 
INFO[0000] args: [-c apk add --no-cache bash]           
INFO[0000] util.Lookup returned: &{Uid:0 Gid:0 Username:root Name:root HomeDir:/root} 
INFO[0000] performing slow lookup of group ids for root 
INFO[0000] Running: [/bin/sh -c apk add --no-cache bash] 
fetch XXX/alpine-generic-remote/v3.16/main/x86_64/APKINDEX.tar.gz
fetch XXX/alpine-generic-remote/v3.16/community/x86_64/APKINDEX.tar.gz
fetch XXX/alpine-generic-remote/edge/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ncurses-terminfo-base (6.3_p20220521-r0)
(2/4) Installing ncurses-libs (6.3_p20220521-r0)
(3/4) Installing readline (8.1.2-r0)
(4/4) Installing bash (5.1.16-r2)
Executing bash-5.1.16-r2.post-install
Executing busybox-1.35.0-r14.trigger
OK: 8 MiB in 19 packages
INFO[0002] Taking snapshot of full filesystem...        
INFO[0002] USER [100](XXX)0                                    
INFO[0002] cmd: USER                                    
INFO[0002] SHELL ["/bin/bash", "-c"]                    
INFO[0002] Pushing image to XXX/debug-kaniko:13327563
```

However, if I try to run the image, there is no bash executable to be found anywhere on the system.
The output of apk under `/lib/apk/db/installed` states:
```
C:Q1owNdUcG0hf9pE5c2MbsYhrwqQGE=
P:bash
V:5.1.16-r2
A:x86_64
S:435099
I:1339392
T:The GNU Bourne Again shell
U:https://www.gnu.org/software/bash/bash.html
L:GPL-3.0-or-later
o:bash
m:Natanael Copa <ncopa@alpinelinux.org>
t:1649795196
c:66da74397b67b6200e7fe9262a14544b0836a1c1
D:/bin/sh so:libc.musl-x86_64.so.1 so:libreadline.so.8
p:cmd:bash=5.1.16-r2
F:bin
R:bash
a:0:0:755
Z:Q1+dNR4F1r7bl/bUBM8Xq3d1LndOk=
...
```

So it looks like it's been installed, but the executable is simply missing.
The `git` package however, which has been installed using the same command, is available and working.

**Expected behavior**
Bash is available in $PATH after installation using apk

**To Reproduce**
1. Copy the Dockerfile given above
2. Build it using Kaniko
3. Verify if bash has been installed by running the image

**Additional Information**
 - Dockerfile 

```Dockerfile
FROM alpine:3.16

USER root

RUN apk add --no-cache bash git

USER 1000

SHELL ["/bin/bash", "-c"]

```
 - Build Context
   None
 - Kaniko Image (fully qualified with digest)
`gcr.io/kaniko-project/executor:debug@sha256:3bc3f3a05f803cac29164ce12617a7be64931748c944f6c419565f500b65e8db`
 
 **Triage Notes for the Maintainers**
 

 
 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | <ul><li>- [ ] </li></ul>|
 | Please check if the build works in docker but not in kaniko | <ul><li>- [ ] </li></ul>| 
 | Please check if this error is seen when you use `--cache` flag | <ul><li>- [ ] </li></ul>|
 | Please check if your dockerfile is a multistage dockerfile | <ul><li>- [ ] </li></ul>|

**Comment by very-doge-wow** [comment]
For anyone else wondering, you can bypass the problem by doing the following in your custom builder image:

```dockerfile
RUN mv /bin/bash /kaniko/bash && ln -sf /kaniko/bash /bin/sh  # symlink is necessary!
SHELL ["/kaniko/bash","-c"]
```

When building with the image resulting from that, bash seems to be properly installed and working in the target image.

**Comment by jaykv** [comment]
> For anyone else wondering, you can bypass the problem by doing the following in your custom builder image:
> 
> ```dockerfile
> RUN mv /bin/bash /kaniko/bash && ln -sf /kaniko/bash /bin/sh  # symlink is necessary!
> SHELL ["/kaniko/bash","-c"]
> ```
> 
> When building with the image resulting from that, bash seems to be properly installed and working in the target image.
> 
> TLDR; You need to put any binaries you **add** to your builder image inside the `/kaniko` dir, because otherwise if you try to install the same packages in your actual image build, kaniko will remove them after the build has completed. So just put the binaries into the kaniko dir (and maybe export your PATH to look into the kaniko dir).

Thanks for the workaround! The binaries in /kaniko were getting ignored by my builder image so I had to put them in /root/bin and add that dir to PATH. This is with kaniko v1.9.0:

builder image:
```Dockerfile
FROM alpine:3.14

...

RUN apk update && apk upgrade --no-cache && apk add --no-cache openssh-client bash
RUN mkdir /root/bin

ENV PATH="$PATH:/root/bin"

RUN mv /usr/bin/bash /root/bin/bash && ln -sf /root/bin/bash /bin/sh; \
    mv /usr/bin/ssh /root/bin/ssh;

ENTRYPOINT ["/root/bin/bash"]
```

target image:
```Dockerfile
FROM node:16.13-alpine3.14

RUN apk update && apk upgrade --no-cache && apk add --no-cache openssh-client git xvfb-run

...
```

**Comment by henricook** [comment]
We've had to employ this workaround, this is a super strange issue no? Like Kaniko is changing the alpine base image itself? Does anyone know more about the cause of this?

**Comment by aaron-prindle** [comment]
In doing some poking around with the following:
```
FROM alpine:3.16

USER root

RUN apk add --no-cache bash git

USER 1000

RUN sh -c "while :; do echo 'just looping here... nothing special'; sleep 1; done"

# SHELL ["/bin/bash", "-c"]
```
and exec-ing into the looping kaniko container:
```
$ docker exec -it b4465126c3f6 sh
```
I see that there is a `bash` binary installed but the location it is installed is not part of the PATH:
```
/ # ls /bin/bash
/bin/bash
/ # which bash
/ # env
USER=root
HOSTNAME=b4465126c3f6
DOCKER_CREDENTIAL_GCR_CONFIG=/kaniko/.config/gcloud/docker_credential_gcr_config.json
SHLVL=1
HOME=/root
OLDPWD=/bin
TERM=xterm
PATH=/usr/local/bin:/kaniko:/busybox
SSL_CERT_DIR=/kaniko/ssl/certs
DOCKER_CONFIG=/kaniko/.docker/
PWD=/
/ # 
```
If anyone in the thread has more context, in this case is the likely root cause then that Kaniko has the incorrect PATH env var, incorredct bash installation directory /bin vs /???, or something else this investigation didn't reveal?  Thanks!

**Comment by aaron-prindle** [comment]
Actually when I try to repro this I am seeing bash installed in the final image:
```
aprindle@aprindle-ssd ~/kaniko  [aprindle-debug]docker run -it gcr.io/aprindle-test-cluster/kaniko-test:latest bash
bash-5.1$ which bash
/bin/bash
bash-5.1$ env
CHARSET=UTF-8
HOSTNAME=1d2a20a04259
PWD=/
HOME=/
LANG=C.UTF-8
TERM=xterm
SHLVL=1
LC_COLLATE=C
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
bash-5.1$ 
```

Can someone describe the exact steps they are using to reproduce missing files?  Thanks

**Comment by very-doge-wow** [comment]
> aaron-prindle

Are you sure you've used kaniko inside a **custom builder image**? As this is the underlying cause, kaniko itself in the officially provided images does not have this issue. However some use-cases need images with more dependencies such as bash as a shell or curl for retrieving build-time dependencies.

**Comment by hq-git** [comment]
In my custom kaniko executor image(base on debian 11 and kaniko v1.15.0).

```dockerfile
## Kaniko basic image
FROM gcr.io/kaniko-project/executor:v1.15.0-debug AS kaniko

## Final basic image
FROM maven:3.8.6-openjdk-8-slim

# Install necessary tools
USER root
COPY ./debian11-sources.list /etc/apt/sources.list
RUN set -eux; \
    apt-get update; \
    apt-get install -y --no-install-recommends ca-certificates curl git git-lfs; \
    apt-get install -y --no-install-recommends openssh-client patch; \
    apt-get clean; \
    rm -rf /var/lib/apt/lists/*

# Install kaniko. Create kaniko directory with world write permission to allow non root run
COPY --from=kaniko /kaniko /kaniko
COPY --from=kaniko /busybox /kaniko_bin
RUN chmod 777 /kaniko

# Declare '/kaniko_bin' as a volume to get it automatically in the path which kaniko to ignore
VOLUME /kaniko_bin
# Move necessary commands to '/kaniko' directory, in case be cleaned by kaniko.
RUN set -eux; \
    mv $(which git) /kaniko_bin/; \
    mv $(which git-lfs) /kaniko_bin/; \
    usermod root -s /kaniko_bin/sh; \
    mkdir -p ${HOME}/.ssh && chmod 700 ${HOME}/.ssh

# Set kaniko env
ENV HOME="/root"
ENV USER="root"
ENV PATH="/kaniko:/kaniko_bin:$PATH"
ENV SSL_CERT_DIR="/kaniko/ssl/certs"
ENV DOCKER_CONFIG="/kaniko/.docker/"
ENV DOCKER_CREDENTIAL_GCR_CONFIG="/kaniko/.config/gcloud/docker_credential_gcr_config.json"

WORKDIR /workspace

ENTRYPOINT [ "/kaniko/executor" ]
CMD [ "--help" ]
```

I found that after execute '/kaniko/executor  xxx --cleanup' to build multi stages image, some folders of rootfs disappeared.

before:

```bash
/workspace # ls -l /
total 92
drwxr-xr-x    1 root     root          4096 Aug  3  2022 bin
drwxr-xr-x    2 root     root          4096 Jun 30  2022 boot
drwxr-xr-x    5 root     root           360 Sep 27 08:10 dev
drwxr-xr-x    1 root     root          4096 Sep 27 08:10 etc
drwxr-xr-x    2 root     root          4096 Jun 30  2022 home
drwxrwxrwx    1 root     root          4096 Aug 29 21:49 kaniko
drwxr-xr-x    2 root     root         12288 Sep 27 08:10 kaniko_bin
drwxr-xr-x    1 root     root          4096 Aug  1  2022 lib
drwxr-xr-x    2 root     root          4096 Aug  1  2022 lib64
drwxr-xr-x    2 root     root          4096 Aug  1  2022 media
drwxr-xr-x    2 root     root          4096 Aug  1  2022 mnt
drwxr-xr-x    2 root     root          4096 Aug  1  2022 opt
dr-xr-xr-x  235 root     root             0 Sep 27 08:10 proc
drwx------    1 root     root          4096 Sep 27 08:10 root
drwxr-xr-x    3 root     root          4096 Aug  1  2022 run
drwxr-xr-x    1 root     root          4096 Aug  3  2022 sbin
drwxr-xr-x    2 root     root          4096 Aug  1  2022 srv
dr-xr-xr-x   13 root     root             0 Sep 27 08:10 sys
drwxrwxrwt    1 root     root          4096 Sep 27 07:47 tmp
drwxr-xr-x    1 root     root          4096 Aug  1  2022 usr
drwxr-xr-x    1 root     root          4096 Aug  1  2022 var
drwxr-xr-x    7 root     root           224 Sep 27 07:47 workspace
```

after:

```bash
/workspace # ls -l /
total 44
drwxr-xr-x    5 0        0              360 Sep 27 07:49 dev
drwxr-xr-x    1 0        0             4096 Sep 27 07:57 etc
drwxrwxrwx    1 0        0             4096 Sep 27 07:57 kaniko
drwxr-xr-x    2 0        0            12288 Sep 27 07:49 kaniko_bin
dr-xr-xr-x  229 0        0                0 Sep 27 07:49 proc
dr-xr-xr-x   13 0        0                0 Sep 27 07:49 sys
drwxrwxrwt    1 0        0             4096 Sep 27 07:57 tmp
drwxr-xr-x    1 0        0             4096 Sep 27 07:57 var
drwxr-xr-x   17 0        0              544 Sep 26 10:57 workspace
```

Even I moved the 'git' command to '/kaniko_bin/', I still got below errors, but the 'patch' command still works. I think it's because 'patch' command is an binary command from kaniko image, it has no external dependency. But for 'git' command, it relies on some  dependencies under rootfs.

```bash
/workspace # which git
/kaniko_bin/git
/workspace # ls -l /kaniko_bin/git
-rwxr-xr-x    1 0        0          3347408 Feb 22  2023 /kaniko_bin/git
/workspace # git -h
/kaniko_bin/sh: git: not found

/workspace # which patch
/kaniko_bin/patch
/workspace # ls -l /kaniko_bin/patch
-rwxr-xr-x  398 0        0          1161032 Jul 17 19:27 /kaniko_bin/patch
/workspace # patch --help
BusyBox v1.36.1 (2023-07-17 19:24:58 UTC) multi-call binary.

Usage: patch [-RNE] [-p N] [-i DIFF] [ORIGFILE [PATCHFILE]]

	-p N	Strip N leading components from file names
	-i DIFF	Read DIFF instead of stdin
	-R	Reverse patch
	-N	Ignore already applied patches
	-E	Remove output files if they become empty
	--dry-run	Don't actually change files
```

Since the '/etc/passwd' is also deleted by kaniko, I can not enter the kaniko container, so I can not do next build with this kaniko container.

```bash
$ docker exec -ti recursing_turing /kaniko_bin/sh
unable to find user root: no matching entries in passwd file
```
---

I know kaniko will extract the files of base image in Dockerfile to current container's rootfs, if I used ADD or COPY in Dockerfile.

```bash
Unpacking rootfs as cmd COPY --from=builder /app/ ./ requires it.
```

Is it doable to have a option like '--backup' to backup rootfs before executing the '/kaniko/executor  xxx --cleanup', and restore the rootfs when build is finished?

**Comment by hq-git** [comment]
Added a script to my custom kaniko image, and it works well. Wish can help guys get out of trouble. 

`kaniko_executor.sh`

```bash
#!/kaniko_bin/sh

set -e

BACKUP_DIR="/kaniko/backup"
# Directories need to backup, the most important ones are "/etc /usr /lib /lib64"
TARGET_DIR="/etc /usr /lib /lib64 /root /sbin /bin /var"

# Backup the important directories under rootfs.
rootfs_backup () {
  echo "-> Backup important directories under rootfs ..."
  if [ ! -d "${BACKUP_DIR}" ]; then mkdir -p "${BACKUP_DIR}"; fi
  for dir_name in ${TARGET_DIR}; do
      if [ -d "${BACKUP_DIR}/${dir_name}" ]; then continue; fi
      cp -aR "${dir_name}" "${BACKUP_DIR}/"
  done
}

# Restore the directories
rootfs_restore () {
  echo "-> Restore important directories under rootfs ..."
  set +e
  for dir_name in $(ls "${BACKUP_DIR}" |xargs); do
      rm -rf "/${dir_name}" 2> /dev/null
      cp -aR "${BACKUP_DIR}/${dir_name}" / 2> /dev/null
  done
  echo "-> Finished to restore rootfs, ready for the next build !"
}

# Execute the kaniko executor command
run_executor () {
  rootfs_backup
  /kaniko/kaniko_executor $*
  rootfs_restore
}

# Main
case "$1" in
    restore)
      rootfs_restore ;;
    *)
      run_executor $* ;;
esac
```

```dockerfile
# Add backup and restore rootfs process.
COPY ./kaniko_executor.sh /kaniko/kaniko_executor.sh
RUN set -eux; \
    chmod +x /kaniko/kaniko_executor.sh; \
    mv /kaniko/executor /kaniko/kaniko_executor; \
    mv kaniko/kaniko_executor.sh /kaniko/executor
```

**Comment by kalavt** [comment]
custom kaniko builder image can not add any packages.
as the same package will not been identified as changed in `Taking snapshot of full filesystem`

**Comment by holmesb** [comment]
I'm seeing this on official Kaniko (not custom) : gcr.io/kaniko-project/executor:v1.23.1-debug
```
FROM alpine:3
RUN apk add --no-cache bash curl
CMD ["/bin/bash"]
```
```
docker run -it <image> bash
Unable to find image '<image>' locally
0.0.1052: Pulling from <image>
43c4264eed91: Already exists
358b32d7664e: Pull complete
Digest: sha256:fb9d5c868634d80be2ff7cc7b6e1570f541374957d9a5be5cc2b9eed6d4fcabf
Status: Downloaded newer image for <image>
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: "bash": executable file not found in $PATH: unknown.
```
Works fine if I docker build.

Command is: `/kaniko/executor --dockerfile Dockerfile --destination <image> --verbosity=debug --context dir://.`  Interestingly, if I follow @aaron-prindle's example and add USER 1000, it works fine. Strange how changing to a non-existent user:
```
bash-5.2$ whoami
whoami: unknown uid 1000
```
should be necessary.

**Comment by very-doge-wow** [comment]
@holmesb Your problem is a different one. As you can tell from the docker run command, you didn't actually start the image you built using kaniko but rather pulled one from Docker Hub, which didn't have bash.

**Comment by holmesb** [comment]
"you didn't actually start the image you built" - `docker run -it <image> bash` 
???
I didn't say which image I started. You'll just have to take my word that it's the image that Kaniko built. Yes I pushed to a registry since the Kaniko executor arguments I gave don't exclude pushing. Can you reproduce? Anyway workarounds seem valid. Just means Kaniko can't simply be slotted-in for docker, Dockerfile change is needed.

**Comment by very-doge-wow** [comment]
@holmesb I was referring to this line in the logoutput:
```Unable to find image '<image>' locally```

Usually a locally built image is cached and needn't be pulled again from the registry for running it.

**Comment by holmesb** [comment]
Wasn't built locally
